metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -1874,8 +1873,13 @@ public CompletableFuture<BrokerHeartbeatReply> processBrokerHeartbeat(
 
                 @Override
                 public ControllerResult<BrokerHeartbeatReply> generateRecordsAndResult() {
+                    Long offsetForRegisterBrokerRecord = clusterControl.registerBrokerRecordOffset(brokerId);
+                    if (offsetForRegisterBrokerRecord == null) {
+                        throw new RuntimeException(
+                            String.format("Receive a heartbeat from broker %d before registration", brokerId));
```

Comment:
I would argue that this means that the client/broker sent a heartbeat request without sending a register broker request first. Unfortunately, KIP-631 only documents `NOT_CONTROLLER` as a possible error code. Looking at the broker code, the broker retries all errors. I say we return an `INVALID_REQUEST` instead of `UNKNOWN_SERVER_ERROR`.

Also, by throwing a `RuntimeException`, this will force the active controller to `resign`. See `QuorumController#handleEventException`. We don't want to do this in this case. For example, this code can instead throw a `InvalidRequestException` and `ControlerApis` will convert that to an `Errors`.

Reply:
Good catch, I find we have a similar inspection at `ClusterControlManager.checkBrokerEpoch` where we return `StaleBrokerEpochException` if BrokerRegistration is null, I also return a `StaleBrokerEpochException` here to make it consistent.